File: 1971-find-if-path-exists-in-graph/find-if-path-exists-in-graph.java
import java.util.*;

class Solution {
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        Map<Integer, List<Integer>> g = new HashMap<>();
        for (int i = 0; i < n; i++) {
            g.put(i, new ArrayList<>());
        }
        for (int[] edge : edges) {
            int u = edge[0];
            int v = edge[1];
            g.get(u).add(v);
            g.get(v).add(u);
        }

        Set<Integer> visited = new HashSet<>();
        Deque<Integer> q = new LinkedList<>();
        q.offerLast(source);

        while (!q.isEmpty()) {
            int curr = q.pollFirst();
            if (curr == destination) {
                return true;
            }
            visited.add(curr);
            for (int i : g.get(curr)) {
                if (!visited.contains(i)) {
                    visited.add(i);
                    q.offerLast(i);
                }
            }
        }

        return false;
    }
}

File: 2121-find-if-path-exists-in-graph/find-if-path-exists-in-graph.py
class Solution(object):
    def validPath(self, n, edges, source, destination):
        g = {i: [] for i in range(n)}
        for u, v in edges:
            g[u].append(v)
            g[v].append(u)

        visited = set()
        q = deque([source])

        while q:
            curr = q.popleft()
            if curr == destination:
                return True
            visited.add(curr)
            for i in g[curr]:
                if i not in visited:
                    visited.add(i)
                    q.append(i)
        
        return False